extensions-core/druid-bloom-filter/src/main/java/org/apache/druid/query/aggregation/bloom/BloomFilterMergeBufferAggregator.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.bloom;
+
+import com.fasterxml.jackson.databind.util.ByteBufferBackedOutputStream;
+import org.apache.druid.io.ByteBufferInputStream;
+import org.apache.druid.query.aggregation.BufferAggregator;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.hive.common.util.BloomKFilter;
+
+import java.nio.ByteBuffer;
+
+public class BloomFilterMergeBufferAggregator implements BufferAggregator
+{
+  private ColumnValueSelector<BloomKFilter> selector;
+  private int maxNumEntries;
+
+  public BloomFilterMergeBufferAggregator(
+      ColumnValueSelector<BloomKFilter> selector,
+      int maxNumEntries
+  )
+  {
+    this.selector = selector;
+    this.maxNumEntries = maxNumEntries;
+  }
+
+  @Override
+  public void init(ByteBuffer buf, int position)
+  {
+    final ByteBuffer mutationBuffer = buf.duplicate();
+    mutationBuffer.position(position);
+    BloomKFilter filter = new BloomKFilter(maxNumEntries);
+    ByteBufferBackedOutputStream outputStream = new ByteBufferBackedOutputStream(mutationBuffer);
+    try {
+      BloomKFilter.serialize(outputStream, filter);
+    }
+    catch (Exception ex) {
+      throw new RuntimeException("Failed to initialize bloomK filter", ex);
+    }
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position)
+  {
+    final int oldPosition = buf.position();
+    final int oldLimit = buf.limit();
+    try {
+      buf.position(position);
+      BloomKFilter collector = BloomKFilter.deserialize(new ByteBufferInputStream(buf));
```

Comment:
It there a way to update Bloom Filter in place without streaming deserialization and than later serialization back?

Reply:
There is a static method, [`BloomKFilter.mergeBloomFilterBytes`](https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hive/common/util/BloomKFilter.java#L360) which can do a merge of 2 bloom filters directly, so this could be done for the merge operation at least, will look into it.

Reply:
Added `ByteBuffer` methods to our copied `BloomKFilter` to add items to and merge serialized bloom filters in place, so buffer aggs should be much more chill to the heap